Client/Frontend/Browser/SearchEngines.swift

```diff
@@ -157,112 +157,55 @@ class SearchEngines {
         NSKeyedArchiver.archiveRootObject(customEngines, toFile: self.customEngineFilePath())
     }
 
-    /// Return all possible paths for a language identifier in the order of most specific to least specific.
-    /// For example, zh-Hans-CN with a default of en will return [zh-Hans-CN, zh-CN, zh, en]. The fallback
-    /// identifier must be a known one that is guaranteed to exist in the SearchPlugins directory.
-    class func directoriesForLanguageIdentifier(_ languageIdentifier: String, basePath: NSString, fallbackIdentifier: String) -> [String] {
-        var directories = [String]()
+    /// Return all possible language identifiers in the order of most specific to least specific.
+    /// For example, zh-Hans-CN will return [zh-Hans-CN, zh-CN, zh].
+    class func possibilitiesForLanguageIdentifier(_ languageIdentifier: String) -> [String] {
+        var possibilities: [String] = []
         let components = languageIdentifier.components(separatedBy: "-")
-        if components.count == 1 {
-            // zh
-            directories.append(languageIdentifier)
-        } else if components.count == 2 {
-            // zh-CN
-            directories.append(languageIdentifier)
-            directories.append(components[0])
-        } else if components.count == 3 {
-            directories.append(languageIdentifier)
-            directories.append(components[0] + "-" + components[2])
-            directories.append(components[0])
-        }
-        if !directories.contains(fallbackIdentifier) {
-            directories.append(fallbackIdentifier)
-        }
-        
-        return directories.map { (path) -> String in
-            return basePath.appendingPathComponent(path)
-        }
-    }
+        possibilities.append(languageIdentifier)
 
-    // Return the language identifier to be used for the search engine selection. This returns the first
-    // identifier from preferredLanguages and takes into account that on iOS 8, zh-Hans-CN is returned as
-    // zh-Hans. In that case it returns the longer form zh-Hans-CN. Same for traditional Chinese.
-    //
-    // These exceptions can go away when we drop iOS 8 or when we start using a better mechanism for search
-    // engine selection that is not based on language identifier.
-    class func languageIdentifierForSearchEngines() -> String {
-        let languageIdentifier = Locale.preferredLanguages.first!
-        switch languageIdentifier {
-            case "zh-Hans":
-                return "zh-Hans-CN"
-            case "zh-Hant":
-                return "zh-Hant-TW"
-            default:
-                return languageIdentifier
+        if components.count == 3, let first = components.first, let last = components.last {
+            possibilities.append("\(first)-\(last)")
+        }
+        if components.count >= 2, let first = components.first {
+            possibilities.append("\(first)")
         }
+        return possibilities
```

Comment:
This code is cleaner than what went before, to be sure. 

```swift
// These exceptions can go away when we drop iOS 8 or when we start using a better mechanism for search
// engine selection that is not based on language identifier.
```
What happens when we drop iOS8? What's in iOS9?

Reply:
Apple changed Locale Identifiers